File: app/src/main/java/com/jameschamberlain/footballteamtracker/fixtures/EditFixtureFragment.java
package com.jameschamberlain.footballteamtracker.fixtures;

import android.app.AlertDialog;
import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.NumberPicker;
import android.widget.TextView;
import android.widget.TimePicker;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.jameschamberlain.footballteamtracker.FileUtils;
import com.jameschamberlain.footballteamtracker.Player;
import com.jameschamberlain.footballteamtracker.R;
import com.jameschamberlain.footballteamtracker.Team;

import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Locale;
import java.util.Objects;

public class EditFixtureFragment extends Fragment {

    private Fixture fixture;

    /**
     * A list of name of the team's players.
     */
    private ArrayList<String> playerNames = new ArrayList<>();

    private View rootView;

    private ArrayAdapter<String> adapter;

    private TextView scoreTextView;

    private int id;

    private EditRecyclerAdapter goalsAdapter;

    private EditRecyclerAdapter assistsAdapter;

    private TextView dateTextView;
    private TextView timeTextView;
    private final Calendar CALENDAR = Calendar.getInstance();

    DatePickerDialog.OnDateSetListener date = new DatePickerDialog.OnDateSetListener() {

        @Override
        public void onDateSet(DatePicker view, int year, int monthOfYear,
                              int dayOfMonth) {
            CALENDAR.set(Calendar.YEAR, year);
            CALENDAR.set(Calendar.MONTH, monthOfYear);
            CALENDAR.set(Calendar.DAY_OF_MONTH, dayOfMonth);
            updateDateLabel();
        }

    };

    TimePickerDialog.OnTimeSetListener time = new TimePickerDialog.OnTimeSetListener() {
        @Override
        public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
            CALENDAR.set(Calendar.HOUR_OF_DAY, hourOfDay);
            CALENDAR.set(Calendar.MINUTE, minute);
            updateTimeLabel();
        }
    };

    EditFixtureFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        Bundle bundle = this.getArguments();
        this.fixture = Objects.requireNonNull(bundle).getParcelable("fixture");
        this.id = Team.getInstance().getFixtures().indexOf(this.fixture);


        rootView = inflater.inflate(R.layout.fragment_fixture_edit, container, false);

        for (Player player : Team.getInstance().getPlayers()) {
            playerNames.add(player.getName());
        }

        adapter = new ArrayAdapter<>(Objects.requireNonNull(getContext()), R.layout.item_player, playerNames);

        // Set the name of the home team.
        TextView homeTeamTextView = rootView.findViewById(R.id.home_team_text_view);
        homeTeamTextView.setText(fixture.getHomeTeam());

        // Set the score of the fixture.
        scoreTextView = rootView.findViewById(R.id.score_text_view);
        scoreTextView.setText(fixture.getScore().toString());

        // Set the name of the away team.
        TextView awayTextView = rootView.findViewById(R.id.away_team_text_view);
        awayTextView.setText(fixture.getAwayTeam());

        // Set the time.
        TextView timeTextView = rootView.findViewById(R.id.time_text_view);
        timeTextView.setText(fixture.getTimeString());

        setupScoreButton();
        setupDate();
        setupTime();
        setupGoals();
        setupAssists();
        setupDeleteButton();
        setupCancelButton();
        setupConfirmButton();

        // Inflate the layout for this fragment
        return rootView;
    }


    private void setupScoreButton() {
    /*
    Add a click listener for the score button so the score can
    be updated.
    An alert will appear with two spinners: one for the home team
    and one for the away team.
     */
        Button updateScoreButton = rootView.findViewById(R.id.update_score_button);
        updateScoreButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder alert = new AlertDialog.Builder(v.getContext());
                alert.setTitle("Update score:");

                LayoutInflater inflater = (LayoutInflater) Objects.requireNonNull(getActivity()).getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                View view = inflater.inflate(R.layout.dialog_score, null);

                final NumberPicker homeScorePicker = view.findViewById(R.id.homeScorePicker);
                homeScorePicker.setMinValue(0);
                homeScorePicker.setMaxValue(20);

                final NumberPicker awayScorePicker = view.findViewById(R.id.awayScorePicker);
                awayScorePicker.setMinValue(0);
                awayScorePicker.setMaxValue(20);

                if (fixture.getResult() == FixtureResult.UNPLAYED) {
                    homeScorePicker.setValue(0);
                    awayScorePicker.setValue(0);
                }
                else {
                    homeScorePicker.setValue(fixture.getScore().getHome());
                    awayScorePicker.setValue(fixture.getScore().getAway());
                }

                alert.setPositiveButton("Confirm", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        int newHomeScore = homeScorePicker.getValue();
                        int newAwayScore = awayScorePicker.getValue();
                        fixture.setScore(new Score(newHomeScore, newAwayScore));
                        scoreTextView.setText(fixture.getScore().toString());
                    }
                });

                alert.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });

                alert.setView(view);
                AlertDialog dialog = alert.create();
                dialog.show();
            }
        });
    }


    private void setupDate() {
        dateTextView = rootView.findViewById(R.id.date_text_view);
        dateTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // COMPLETE
                new DatePickerDialog(getContext(), date,
                        fixture.getDateTime().getYear(),
                        (fixture.getDateTime().getMonthValue() - 1),
                        fixture.getDateTime().getDayOfMonth())
                        .show();
            }
        });
        CALENDAR.set(Calendar.YEAR, fixture.getDateTime().getYear());
        CALENDAR.set(Calendar.MONTH, (fixture.getDateTime().getMonthValue() - 1));
        CALENDAR.set(Calendar.DAY_OF_MONTH, fixture.getDateTime().getDayOfMonth());

        updateDateLabel();
    }

    private void updateDateLabel() {
        String myFormat = "E, d MMM yyyy";
        SimpleDateFormat sdf = new SimpleDateFormat(myFormat, Locale.UK);
        dateTextView.setText(sdf.format(CALENDAR.getTime()));
    }


    private void setupTime() {
        timeTextView = rootView.findViewById(R.id.time_text_view);
        timeTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new TimePickerDialog(getContext(), time,
                        fixture.getDateTime().getHour(),
                        fixture.getDateTime().getMinute(),
                        true).show();
            }
        });

        CALENDAR.set(Calendar.HOUR_OF_DAY, fixture.getDateTime().getHour());
        CALENDAR.set(Calendar.MINUTE, fixture.getDateTime().getMinute());

        updateTimeLabel();
    }

    private void updateTimeLabel() {
        String myFormat = "kk:mm";
        SimpleDateFormat sdf = new SimpleDateFormat(myFormat, Locale.UK);
        timeTextView.setText(sdf.format(CALENDAR.getTime()));
    }


    private void setupGoals() {

        ArrayList<String> goalscorers = fixture.getGoalscorers();

        // Create an {@link TeamAdapter}, whose data source is a list of {@link Player}s. The
        // adapter knows how to create list items for each item in the list.
        goalsAdapter = new EditRecyclerAdapter(fixture, true);

        // Find the {@link ListView} object in the view hierarchy of the {@link Activity}.
        // There should be a {@link ListView} with the view ID called list, which is declared in the
        // word_list layout file.
        final RecyclerView recyclerView = rootView.findViewById(R.id.goals_list);


        // Make the {@link ListView} use the {@link TeamAdapter} we created above, so that the
        // {@link ListView} will display list items for each {@link Fixture} in the list.
        recyclerView.setAdapter(goalsAdapter);

        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        /*
        Add a click listener for the goalscorers button so the goalscorers
        can be updated.
        An alert will appear with a list of the team's players.
         */
        Button addGoalscorerButton = rootView.findViewById(R.id.add_goalscorer_button);
        addGoalscorerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder alert = new AlertDialog.Builder(v.getContext());
                alert.setTitle("Add a goalscorer:");

                alert.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });

                alert.setAdapter(adapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String name = adapter.getItem(which);
                        ArrayList<String> newGoalscorers = fixture.getGoalscorers();
                        newGoalscorers.add(name);
                        fixture.setGoalscorers(newGoalscorers);
                        goalsAdapter.notifyDataSetChanged();
                    }
                });

                AlertDialog dialog = alert.create();
                dialog.show();
            }
        });
    }


    private void setupAssists() {

        ArrayList<String> assists = fixture.getAssists();

        // Create an {@link TeamAdapter}, whose data source is a list of {@link Player}s. The
        // adapter knows how to create list items for each item in the list.
        assistsAdapter = new EditRecyclerAdapter(fixture, false);

        // Find the {@link ListView} object in the view hierarchy of the {@link Activity}.
        // There should be a {@link ListView} with the view ID called list, which is declared in the
        // word_list layout file.
        final RecyclerView recyclerView = rootView.findViewById(R.id.assists_list);


        // Make the {@link ListView} use the {@link TeamAdapter} we created above, so that the
        // {@link ListView} will display list items for each {@link Fixture} in the list.
        recyclerView.setAdapter(assistsAdapter);

        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));


        /*
        Add a click listener for the assists button so the assists
        can be updated.
        An alert will appear with a list of the team's players.
         */
        Button addAssistButton = rootView.findViewById(R.id.add_assist_button);
        addAssistButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder alert = new AlertDialog.Builder(v.getContext());
                alert.setTitle("Add an assist:");

                alert.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });

                alert.setAdapter(adapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String name = adapter.getItem(which);
                        //addNewView(assistsLayout, name);
                        ArrayList<String> newAssists = fixture.getAssists();
                        newAssists.add(name);
                        fixture.setAssists(newAssists);
                        assistsAdapter.notifyDataSetChanged();
                    }
                });

                AlertDialog dialog = alert.create();
                dialog.show();
            }
        });
    }


    private void setupDeleteButton() {
        Button deleteButton = rootView.findViewById(R.id.delete_button);
        deleteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder alert = new AlertDialog.Builder(v.getContext());
                alert.setTitle("Delete fixture?")
                        .setMessage("Are you sure you would like to delete this fixture?")
                        .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                ArrayList<Fixture> fixtures = Team.getInstance().getFixtures();
                                fixtures.remove(id);
                                // Sort fixtures.
                                Collections.sort(fixtures);
                                // Write the update to a file.
                                FileUtils.writeFixturesFile(fixtures);
                                FragmentManager fm = Objects.requireNonNull(getActivity()).getSupportFragmentManager();
                                if (fm.getBackStackEntryCount() > 0) {
                                    fm.popBackStackImmediate();
                                    if (fm.getBackStackEntryCount() > 0) {
                                        fm.popBackStackImmediate();
                                    }
                                }
                            }
                        })
                        .setNegativeButton("No", null);
                AlertDialog dialog = alert.create();
                dialog.show();
            }
        });
    }


    private void setupConfirmButton() {
        Button confirmButton = rootView.findViewById(R.id.confirm_button);
        confirmButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Update fixture.
                Team team = Team.getInstance();
                Fixture originalFixture = team.getFixtures().get(id);
                // Set the score.
                originalFixture.setScore(fixture.getScore());
                // Get date and time
                int year = CALENDAR.get(Calendar.YEAR);
                int month = CALENDAR.get(Calendar.MONTH) + 1;
                int day = CALENDAR.get(Calendar.DAY_OF_MONTH);
                int hour = CALENDAR.get(Calendar.HOUR_OF_DAY);
                int minute = CALENDAR.get(Calendar.MINUTE);
                originalFixture.setDateTime(LocalDateTime.of(year, month, day, hour, minute));
                // Sort & set goalscorers.
                Collections.sort(fixture.getGoalscorers());
                originalFixture.setGoalscorers(fixture.getGoalscorers());
                // Sort & set assists.
                Collections.sort(fixture.getAssists());
                originalFixture.setAssists(fixture.getAssists());
                // Sort fixtures.
                Collections.sort(team.getFixtures());
                // Update the team's stats.
                team.updateTeamStats();
                // Update the player's stats.
                team.updatePlayerStats();
                // Write the update to a file.
                FileUtils.writeFixturesFile(team.getFixtures());
                // Return to the previous screen.
                FragmentManager fm = Objects.requireNonNull(getActivity()).getSupportFragmentManager();
                if (fm.getBackStackEntryCount() > 0) {
                    fm.popBackStack();
                }
            }
        });
    }


    private void setupCancelButton() {
        Button cancelButton = rootView.findViewById(R.id.cancel_button);
        cancelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm = Objects.requireNonNull(getActivity()).getSupportFragmentManager();
                if (fm.getBackStackEntryCount() > 0) {
                    fm.popBackStackImmediate();
                }
            }
        });
    }
}

File: app/src/main/java/com/jameschamberlain/footballteamtracker/fixtures/EditFixtureFragment.kt
package com.jameschamberlain.footballteamtracker.fixtures

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.SetOptions
import com.jameschamberlain.footballteamtracker.R
import com.jameschamberlain.footballteamtracker.Utils
import com.jameschamberlain.footballteamtracker.adapters.EditFixtureStatAdapter
import com.jameschamberlain.footballteamtracker.data.Fixture
import com.jameschamberlain.footballteamtracker.data.FixtureResult
import com.jameschamberlain.footballteamtracker.data.Score
import com.jameschamberlain.footballteamtracker.databinding.FragmentFixtureEditBinding
import com.jameschamberlain.footballteamtracker.viewmodels.FixturesViewModel
import com.jameschamberlain.footballteamtracker.viewmodels.FixturesViewModelFactory
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.collections.ArrayList

private const val TAG = "EditFixtureFragment"

class EditFixtureFragment internal constructor() : Fragment() {


    private var _binding: FragmentFixtureEditBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    private lateinit var editedFixture: Fixture
    private lateinit var originalFixture: Fixture

    private lateinit var fixtureId: String

    private val calendar = Calendar.getInstance()

    private var playerNames = ArrayList<String>()

    private val model: FixturesViewModel by activityViewModels { FixturesViewModelFactory(Utils.getTeamReference(requireActivity())) }

    private val args: EditFixtureFragmentArgs by navArgs()


    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        _binding = FragmentFixtureEditBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        fixtureId = args.fixtureId

        originalFixture = model.getSelectedFixture().value!!
        editedFixture = originalFixture.copyOf()
        setupUi()

        calendar.timeInMillis = originalFixture.dateTime

    }

    private fun setupUi() {
        binding.apply {
            homeTeamTextView.text = if (editedFixture.isHomeGame) Utils.getTeamName(requireActivity()) else editedFixture.opponent
            awayTeamTextView.text = if (editedFixture.isHomeGame) editedFixture.opponent else Utils.getTeamName(requireActivity())
            scoreTextView.text = editedFixture.score.toString()
        }

        calendar.timeInMillis = editedFixture.dateTime

        setupScoreButton()
        setupDate()
        setupTime()
        setupVenue()

        Utils.getTeamReference(requireActivity()).collection("players").get()
                .addOnSuccessListener {
                    it.documents.forEach { doc ->
                        playerNames.add(doc.getString("name")!!)
                    }
                    val adapter: ArrayAdapter<String> = ArrayAdapter(requireContext(), R.layout.item_player, playerNames)
                    setupGoals(adapter)
                    setupAssists(adapter)
                }
                .addOnFailureListener { e ->
                    Log.e(TAG, "Get failed with ", e)
                }

        setupCancelButton()
        setupConfirmButton()
    }

    private fun setupScoreButton() {
        /*
    Add a click listener for the score button so the score can
    be updated.
    An alert will appear with two spinners: one for the home team
    and one for the away team.
     */
        binding.updateScoreButton.setOnClickListener { v ->
            val view = layoutInflater.inflate(R.layout.dialog_score, null)
            val homeScorePicker = view.findViewById<NumberPicker>(R.id.home_score_picker).apply {
                minValue = 0
                maxValue = 20
                value = if (editedFixture.result == FixtureResult.UNPLAYED) 0 else editedFixture.score.home
            }
            val awayScorePicker = view.findViewById<NumberPicker>(R.id.away_score_picker).apply {
                minValue = 0
                maxValue = 20
                value = if (editedFixture.result == FixtureResult.UNPLAYED) 0 else editedFixture.score.away
            }
            MaterialAlertDialogBuilder(v.context)
                    .setTitle("Update score:")
                    .setPositiveButton(getString(R.string.confirm)) { _, _ ->
                        val newHomeScore = homeScorePicker.value
                        val newAwayScore = awayScorePicker.value
                        editedFixture.score = Score(newHomeScore, newAwayScore)
                        binding.scoreTextView.text = editedFixture.score.toString()
                    }
                    .setNegativeButton(getString(R.string.cancel)) { dialog, _ -> dialog.dismiss() }
                    .setView(view)
                    .show()
        }
    }

    private fun setupDate() {
        val datePicker =
            MaterialDatePicker.Builder.datePicker()
                .setTitleText("Select date")
                .setSelection(calendar.timeInMillis)
                .build()

        datePicker.addOnPositiveButtonClickListener {
            val hour = calendar[Calendar.HOUR_OF_DAY]
            val minute = calendar[Calendar.MINUTE]
            calendar.timeInMillis = it
            calendar[Calendar.HOUR_OF_DAY] = hour
            calendar[Calendar.MINUTE] = minute
            updateDateLabel()
        }
        binding.dateTextView.setOnClickListener {
            datePicker.show(requireActivity().supportFragmentManager, "datePicker")
        }
        updateDateLabel()
    }

    private fun updateDateLabel() {
        val formatter = DateTimeFormatter.ofPattern("E, d MMM yyyy")
        binding.dateTextView.text = formatter.format(LocalDateTime.ofInstant(calendar.toInstant(), calendar.timeZone.toZoneId()))
    }

    private fun setupTime() {
        val timePicker =
            MaterialTimePicker.Builder()
                .setTimeFormat(TimeFormat.CLOCK_24H)
                .setTitleText("Select time")
                .setHour(calendar[Calendar.HOUR_OF_DAY])
                .setMinute(calendar[Calendar.MINUTE])
                .build()

        timePicker.addOnPositiveButtonClickListener {
            calendar[Calendar.HOUR_OF_DAY] = timePicker.hour
            calendar[Calendar.MINUTE] = timePicker.minute
            updateTimeLabel()
        }
        binding.timeTextView.setOnClickListener {
            timePicker.show(requireActivity().supportFragmentManager, "timePicker")
        }
        updateTimeLabel()
    }


    private fun setupVenue() {
        binding.venueTextView.text = if (originalFixture.venue != null) originalFixture.venue else "N/A"
    }

    private fun updateTimeLabel() {
        val formatter = DateTimeFormatter.ofPattern("kk:mm")
        binding.timeTextView.text = formatter.format(LocalDateTime.ofInstant(calendar.toInstant(), calendar.timeZone.toZoneId()))
    }

    private fun setupGoals(adapter: ArrayAdapter<String>) {
        val goalsAdapter = EditFixtureStatAdapter(editedFixture, true)
        binding.goalsRecyclerView.adapter = goalsAdapter
        binding.goalsRecyclerView.layoutManager = LinearLayoutManager(activity)

        /*
        Add a click listener for the goalscorers button so the goalscorers
        can be updated.
        An alert will appear with a list of the team's players.
         */
        val items = Array(playerNames.size) { "" }
        for (x in 0 until playerNames.size) {
            items[x] = playerNames[x]
        }
        binding.addGoalscorerButton.setOnClickListener { v ->
            MaterialAlertDialogBuilder(v.context)
                    .setTitle("Add a goalscorer:")
                    .setItems(items) { _, which ->
                        val name: String = adapter.getItem(which)!!
                        val newGoalscorers = editedFixture.goalscorers
                        newGoalscorers.add(name)
                        editedFixture.goalscorers = newGoalscorers
                        goalsAdapter.notifyDataSetChanged()
                    }
                    .show()
        }
    }

    private fun setupAssists(adapter: ArrayAdapter<String>) {
        val assistsAdapter = EditFixtureStatAdapter(editedFixture, false)
        binding.assistsRecyclerView.adapter = assistsAdapter
        binding.assistsRecyclerView.layoutManager = LinearLayoutManager(activity)

        /*
        Add a click listener for the assists button so the assists
        can be updated.
        An alert will appear with a list of the team's players.
         */
        val items = Array(playerNames.size) { "" }
        for (x in 0 until playerNames.size) {
            items[x] = playerNames[x]
        }
        binding.addAssistButton.setOnClickListener { v ->
            MaterialAlertDialogBuilder(v.context)
                    .setTitle("Add an assist:")
                    .setItems(items) { _, which ->
                        val name: String = adapter.getItem(which)!!
                        val newAssists = editedFixture.assists
                        newAssists.add(name)
                        editedFixture.assists = newAssists
                        assistsAdapter.notifyDataSetChanged()
                    }
                    .show()
        }
    }


    private fun updatePlayersStats() {
        playerNames.forEach { player ->
            val goalChange =
                (Collections.frequency(editedFixture.goalscorers, player) -
                        Collections.frequency(originalFixture.goalscorers, player)).toDouble()
            val assistChange =
                (Collections.frequency(editedFixture.assists, player) -
                        Collections.frequency(originalFixture.assists, player)).toDouble()

            if (goalChange != 0.0 || assistChange != 0.0)
                Utils.getTeamReference(requireActivity()).collection("players").document(player.lowercase(Locale.ROOT))
                    .update(mapOf(
                        "goals" to FieldValue.increment(goalChange),
                        "assists" to FieldValue.increment(assistChange)
                    ))
        }
    }

    private fun setupConfirmButton() {
        binding.confirmButton.setOnClickListener {
            editedFixture.goalscorers.sort()
            editedFixture.assists.sort()
            editedFixture.dateTime = calendar.timeInMillis

            updatePlayersStats()

            Utils.getTeamReference(requireActivity()).collection("fixtures").document(fixtureId)
                    .set(editedFixture, SetOptions.merge())
                    .addOnSuccessListener {
                        Log.d(TAG, "DocumentSnapshot successfully updated!")
                        model.selectFixture(editedFixture)
                        NavHostFragment.findNavController(this@EditFixtureFragment).navigateUp()
                    }
                    .addOnFailureListener { e ->
                        Log.w(TAG, "Error deleting document", e)
                        Toast.makeText(activity, "Failed to edit fixture, try again", Toast.LENGTH_SHORT).show()
                    }
        }
    }

    private fun setupCancelButton() {
        binding.cancelButton.setOnClickListener {
            NavHostFragment.findNavController(this@EditFixtureFragment).navigateUp()
        }
    }


    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}